atmel-samd/usb.c

```diff
@@ -267,18 +273,35 @@ int usb_read(void) {
     return data;
 }
 
+// TODO(tannewt): See if we can disable the internal CDC IN cache since we
+// we manage this one ourselves.
+COMPILER_ALIGNED(4) uint8_t cdc_output_buffer[64];
+
 void usb_write(const char* buffer, uint32_t len) {
     if (!cdc_enabled()) {
         return;
     }
-    int32_t result = cdcdf_acm_write((uint8_t *)buffer, len);
-    while (result == USB_BUSY) {
-        #ifdef MICROPY_VM_HOOK_LOOP
-            MICROPY_VM_HOOK_LOOP
-        #endif
-        result = cdcdf_acm_write((uint8_t *)buffer, len);
+    uint8_t * output_buffer;
+    uint8_t output_len;
+    while (len > 0) {
+        while (usb_transmitting) {}
+        output_buffer = (uint8_t *) buffer;
+        output_len = len;
+        // Use our own cache in two different cases:
+        //   * When we're at the end of a transmission and we'll return before
+        //     the given buffer is actually transferred to the USB device.
+        //   * When our given buffer isn't aligned on word boundaries.
+        if (output_len <= 64 || ((uint32_t) buffer) % 4 != 0) {
+            output_buffer = cdc_output_buffer;
+            memcpy(cdc_output_buffer, buffer, output_len);
```

Comment:
If `len` (same as `output_len` here) > 64 AND the input `buffer` is not aligned on a 4-byte boundary, then I think the `memcpy` will write past the end of `cdc_output_buffer`, which is only 64 bytes long.

Reply:
Did you mean `&&`  ??

Reply:
Nah, if we're on the last packet we need to copy to the buffer because we'll return before its transmitted. I added a length check.

Reply:
but if the buffer is "misaligned" don't you have to do this on every transmission? As Dan pointed out, if buffer length is > 64 and misaligned it will send it all on the first pass.

Reply:
Check the latest version. :-)

Reply:
oops!